src/main/java/io/fabric8/jenkins/openshiftsync/BuildSyncRunListener.java

```diff
@@ -0,0 +1,380 @@
+  public static final String ANNOTATION_JENKINS_BUILD_URL = "openshift.io/jenkins-build-uri";
+  public static final String ANNOTATION_JENKINS_STATUS_JSON = "openshift.io/jenkins-status-json";
+  public static final String ANNOTATION_OPENSHIFT_BUILD_NUMBER = "openshift.io/build.number";
+
+  public static final String LABEL_BUILDCONFIG = "buildconfig";
+  public static final String LABEL_OPENSHIFT_BUILD_CONFIG_NAME = "openshift.io/build-config.name";
+
+  @XStreamOmitField
+  private final Logger logger = Logger.getLogger(getClass().getName());
+
+  private long pollPeriodMs = 2000;
+  private String server;
+  private String namespace;
+
+  @XStreamOmitField
+  private OpenShiftClient openShiftClient;
+
+  @XStreamOmitField
+  private Set<String> urlsToPoll = new CopyOnWriteArraySet<>();
+  @XStreamOmitField
+  private AtomicBoolean timerStarted = new AtomicBoolean(false);
+
+  public BuildSyncRunListener() {
+    openShiftClient = init();
+  }
+
+  @DataBoundConstructor
+  public BuildSyncRunListener(String server, String namespace, long pollPeriodMs) {
+    this.server = server;
+    this.namespace = namespace;
+    this.pollPeriodMs = pollPeriodMs;
+  }
+
+  private OpenShiftClient init() {
+    OpenShiftClient openShiftClient;
+    openShiftClient = OpenShiftUtils.createOpenShiftClient(server);
+    return openShiftClient;
+  }
+
+  @Override
+  public void onStarted(Run run, TaskListener listener) {
+    String url = run.getUrl();
+    if (shouldPollRun(run)) {
+      checkTimerStarted();
+      if (urlsToPoll.add(url)) {
+        logger.info("starting polling build " + url);
+      }
+    }
+    super.onStarted(run, listener);
+  }
+
+  protected void checkTimerStarted() {
+    if (timerStarted.compareAndSet(false, true)) {
+      Runnable task = new Runnable() {
+        @Override
+        public void run() {
+          pollLoop();
+        }
+      };
+      Timer.get().scheduleAtFixedRate(task, pollPeriodMs, pollPeriodMs, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  @Override
+  public void onCompleted(Run run, @Nonnull TaskListener listener) {
+    String url = run.getUrl();
+    urlsToPoll.remove(url);
+    pollRun(url, run);
+    logger.info("onCompleted " + url);
+    super.onCompleted(run, listener);
+  }
+
+  @Override
+  public void onDeleted(Run run) {
+    String url = run.getUrl();
+    urlsToPoll.remove(url);
+    pollRun(url, run);
+    logger.info("onDeleted " + url);
+    super.onDeleted(run);
+
+    // TODO should we remove the OpenShift Build too?
+  }
+
+  @Override
+  public void onFinalized(Run run) {
+    String url = run.getUrl();
+    urlsToPoll.remove(url);
+    pollRun(url, run);
+    logger.info("onFinalized " + url);
+    super.onFinalized(run);
+  }
+
+
+  protected void pollLoop() {
+    for (String url : urlsToPoll) {
+      pollRun(url, null);
+    }
+  }
+
+
+  protected void pollRun(String url, Run run) {
+    BuildName buildName = BuildName.parseBuildUrl(url);
+
+    String root = JenkinsUtils.getRootUrl();
```

Comment:
Lets see how easy it is - raised this issue to track:
https://github.com/fabric8io/openshift-jenkins-sync-plugin/issues/12

would love to figure out a way to get notified on a Stage completing too; to avoid all polling!